pytorch_lightning/trainer/trainer.py

```diff
@@ -500,14 +509,16 @@ def _call_and_handle_interrupt(self, trainer_fn, *args, **kwargs):
                 self.state.status = TrainerStatus.INTERRUPTED
                 self.on_keyboard_interrupt()
                 # same treatment as below
-                self.accelerator.teardown()
+                if not isinstance(self.training_type_plugin, DDPSpawnPlugin):
+                    self.accelerator.teardown()
         except BaseException:
             self.state.status = TrainerStatus.INTERRUPTED
             if distributed_available() and self.world_size > 1:
                 # try syncing remaing processes, kill otherwise
                 self.training_type_plugin.reconciliate_processes(traceback.format_exc())
             # give accelerators a chance to finish
-            self.accelerator.teardown()
+            if not isinstance(self.training_type_plugin, DDPSpawnPlugin):
+                self.accelerator.teardown()
```

Comment:
@awaelchli another area where spawning comes in. @daniellepintz pointed out that without this check, `self.lightning_module` here will be None in the main process: https://github.com/PyTorchLightning/pytorch-lightning/blob/24f0124ddd478e18b76d64a79f1d7b341a0f8646/pytorch_lightning/plugins/training_type/parallel.py#L141-L143

which feels a lot like https://github.com/PyTorchLightning/pytorch-lightning/pull/8685

Causing the teardown to fail. 
My questions: 
- Do we really need to call accelerator.teardown() in the case of this exception? What was the purpose in calling accelerator.on_train_end earlier? 
- Why only the accelerator's teardown and not the profiler, logger, or other components? 
- Do we need special error handling for spawn plugins?

Reply:
> Do we really need to call accelerator.teardown() in the case of this exception? What was the purpose in calling accelerator.on_train_end earlier?
> Why only the accelerator's teardown and not the profiler, logger, or other components?

accelerator teardown is currently not called in the exception handling here, only on post_dispatch on a successful completion of the trainer function. The question is should we? Some accelerators try to move the model or call a barrier, this could obviously fail due to chain reactions, as we already are in the exception handling at this point. 

accelerator.on_train_end is currently only implemented by the ipu accelerator/plugin. We could look for an alternative place, I'm not sure if it's needed or not. Maybe @SeanNaren will know.

the only other place where on_train_end is called is in the fit loop, on run end.

Reply:
I personally don't think we should because of those chain reactions. exceptions during exception handling will mask errors from users, which will be really bad .